### DISCUSSION TITLE
byte/string representation query

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#9765] [Using MicroPython] [answered]
Poking around in a .mpy file I noticed it's entrails display differently depending on how I open the file. 
If I open it 'rb' as a bytes file the beginning prints as b"M\x05\x02\x1f \xc6(\xd4\x02\xc6\x04
but if I open it 'r' as a string, file there are differences.   "M\x05\x02\x1f \x06(\x14\x02\x06\x04
I don't really understand why. 
I ask this question with some trepidation after a couple of replies on this forums predecessor that ranged from agro to dismissive. Please remember we're all different distances into the coding journey, so please be patient & reasonable if you feel I'm asking about the obvious.

**Comment by dlech** [comment, marked as answer]
The .mpy file format is a binary file format, so it really only makes sense to open it using the binary ('rb') option. Otherwise, it opens it as a text file and decodes the contents as a UTF-8 string object (which could actually cause an error if the file contains invalid UTF-8 sequences).

When opened as a binary file, Python reads the contents of the file as a [bytes](https://docs.python.org/3.8/library/stdtypes.html#bytes) object instead of a string object. The way Python prints a bytes object looks very much like a string - there is just a `b` before the opening `"`.

**Reply by kjm1102** [reply]
I want to understand how some bit patterns (like 0001 1111) look the same as a byte object or a string object (1f) but the objects representing other bit patterns (eg 1100 0110) are different (c6 for a bytes object but 06 for a string object)?

**Reply by dlech** [reply]
Because the string is the data interpreted/decoded as UTF-8, not individual bytes. UTF-8 can contain multi-byte characters. https://en.wikipedia.org/wiki/UTF-8 shows the bit patterns that will trigger this.

**Reply by kjm1102** [reply]
Ah yes, I get it now, tnx David.
